core/src/main/java/org/jdbi/v3/core/mapper/GenericMapMapperFactory.java

```diff
@@ -0,0 +1,127 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.core.mapper;
+
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.sql.ResultSet;
+import java.sql.ResultSetMetaData;
+import java.sql.SQLException;
+import java.text.MessageFormat;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.UnaryOperator;
+import org.jdbi.v3.core.config.ConfigRegistry;
+import org.jdbi.v3.core.generic.GenericType;
+import org.jdbi.v3.core.generic.GenericTypes;
+import org.jdbi.v3.core.statement.StatementContext;
+import org.jdbi.v3.meta.Beta;
+
+@Beta
+// TODO jdbi4: consolidate with MapMapper and include by default
+public class GenericMapMapperFactory implements RowMapperFactory {
+    // invoked by sqlobject
+    @Override
+    public Optional<RowMapper<?>> build(Type mapType, ConfigRegistry config) {
+        return Optional.of(mapType)
+            .filter(ParameterizedType.class::isInstance)
+            .map(ParameterizedType.class::cast)
+            .filter(maybeMap -> Map.class.equals(maybeMap.getRawType()))
+            .filter(map -> String.class.equals(GenericTypes.findGenericParameter(map, Map.class, 0).orElse(null)))
+            .flatMap(map -> GenericTypes.findGenericParameter(map, Map.class, 1))
+            .filter(value -> !Object.class.equals(value))
+            .flatMap(config.get(ColumnMappers.class)::findFor)
+            .map(GenericMapMapper::new);
+    }
+
+    // invoked manually or by fluent api
+    @SuppressWarnings("unchecked")
+    @Beta
+    public static <T> GenericMapMapper<T> getMapperForValueType(Class<T> valueType, ConfigRegistry config) {
+        return (GenericMapMapper<T>) getMapperForValueType((Type) valueType, config);
+    }
+
+    // invoked manually or by fluent api
+    @SuppressWarnings("unchecked")
+    @Beta
+    public static <T> GenericMapMapper<T> getMapperForValueType(GenericType<T> valueType, ConfigRegistry config) {
+        return (GenericMapMapper<T>) getMapperForValueType(valueType.getType(), config);
+    }
+
+    private static GenericMapMapper<?> getMapperForValueType(Type valueType, ConfigRegistry config) {
+        return config.get(ColumnMappers.class)
+            .findFor(valueType)
+            .map(GenericMapMapper::wildcard)
+            .orElseThrow(() -> new RuntimeException("no column mapper found for type " + valueType));
+    }
+
+    private static class GenericMapMapper<T> implements RowMapper<Map<String, T>> {
+        private final ColumnMapper<T> mapper;
+
+        private GenericMapMapper(ColumnMapper<T> mapper) {
+            this.mapper = mapper;
+        }
+
+        private static GenericMapMapper<?> wildcard(ColumnMapper<?> mapper) {
```

Comment:
This is a bit of an oddball that seems to exist only to *throw away* generic information.  How come it's not `<T> GMM<T> of(CM<T> mapper)`?

Reply:
![image](https://user-images.githubusercontent.com/3950300/49309175-a00c2b00-f4da-11e8-86a6-b11b268d2892.png)

It was the cleanest solution possible. I spent a lot of time shuffling code and parameters around yesterday, this is the only form I was acceptably happy with (the fewest ugly hacks and fake generics).

Reply:
I don't find `.map((Function<ColumnMapper<?>, ? extends GenericMapMapper<?>>) GenericMapMapper::new)` any better tbh

Reply:
After adding that in, I inlined the intermediate factory method (turning it into 2 different `findFor` calls again), and with that I can all of a sudden use the constructor instead of the factory method again, and basically all casts and stuff are gone.

Much confusion. ¯\\\_(ツ)\_/¯

Reply:
Well I mean, the final result now makes perfect sense, but I wonder why it gave me so many errors along the way, half of which didn't make much sense, that made me write so much roundabout code in the first place...